### PR TITLE
Use trust bundle to authenticate TLS connection

### PR DESCRIPTION
**Pull request by thunderboltsid**
AdditionalTrustBundle stores a PEM-encoded certificate chain for the underlying Prism Central.

**Note**: Clusterctl uses envsubst (https://github.com/drone/envsubst) to replace variables in the templates. This is	problematic for multi-line content like the cert bundle. envsubst will replace the variables, honor the newlines, but ignore the expected indentation in the resulting YAML file. The workaround is to set the environment variable `NUTANIX_ADDITIONAL_TRUST_BUNDLE` to a one-liner cert generated using a `base64` encoding.
```
cat ca-cert.pem | base64
``` 

The resulting ConfigMap for the trust bundle then stores the base64 encoded value as BinaryData. Prism Go Client's kubernetes environment provider can hydrade the CA cert from binary data. 

**How was this tested?**
Run `LABEL_FILTERS="capx-feature-test"  make test-e2e-calico` against a Prism Cluster (@vnephologist lab) with Self-Signed certificates that have the IP Address populated in SAN.

```
Ran 12 of 23 Specs in 1758.240 seconds
SUCCESS! -- 12 Passed | 0 Failed | 0 Pending | 11 Skipped
--- PASS: TestE2E (1758.25s)
PASS

Ginkgo ran 1 suite in 29m27.049073463s
```

### CODE REVIEW

File: controllers/nutanixcluster_controller.go [hunk 1]
```diff
@@ -50,9 +50,10 @@ import (
 
 // NutanixClusterReconciler reconciles a NutanixCluster object
 type NutanixClusterReconciler struct {
-	Client         client.Client
-	SecretInformer coreinformers.SecretInformer
-	Scheme         *runtime.Scheme
+	Client            client.Client
+	SecretInformer    coreinformers.SecretInformer
+	ConfigMapInformer coreinformers.ConfigMapInformer
```
Comment (deepakm-ntnx) [inline controllers/nutanixcluster_controller.go:55]: its unclear to me why this new dependency on cmInformer is required?

**Review by deepakm-ntnx — COMMENTED** [review]
COuld you please post the test status in description and add tests as well

**Comment by deepakm-ntnx** [comment]
added some comments on https://github.com/nutanix-cloud-native/prism-go-client/pull/37 as well

File: controllers/nutanixcluster_controller.go (hunk 1 above)
Comment (thunderboltsid) [reply controllers/nutanixcluster_controller.go:55]: To fetch the trust bundle from the configmap

**Comment by thunderboltsid** [comment]
> COuld you please post the test status in the description and add tests as well

Once I have run the e2e tests successfully, yes. Currently, the PR is still WIP.

**Comment by thunderboltsid** [comment]
Currently blocked on testing the changes as the only cluster I had access to that had proper self-signed certificates was @vnephologist  lab that seems to have started disconnecting my VPN connection attempts.

File: test/e2e/test_helpers.go [hunk 2]
```diff
@@ -279,6 +279,8 @@ func (t testHelper) deployClusterAndWait(params deployClusterParams, clusterReso
 		ControlPlaneMachineCount: pointer.Int64Ptr(1),
 		WorkerMachineCount:       pointer.Int64Ptr(1),
 	}
+	fmt.Println("apparently Custer config")
```
Comment (deepakm-ntnx) [inline test/e2e/test_helpers.go:282]: are these logs useful? also nit: spelling of cluster is wrong.

Comment (thunderboltsid) [reply test/e2e/test_helpers.go:282]: nope; just debug logs. will be cleaned up.

**Comment by thunderboltsid** [comment]
CSI e2e is failing due to missing env vars (we should update test/e2e/README.md)
```
Nutanix flavor CSI
  Create a cluster with Nutanix CSI and use Nutanix Volumes to create PV [capx-feature-test, csi, slow, network]
  /Users/sid.shukla/go/src/github.com/nutanix-cloud-native/cluster-api-provider-nutanix/test/e2e/csi_test.go:101
[BeforeEach] Nutanix flavor CSI
  /Users/sid.shukla/go/src/github.com/nutanix-cloud-native/cluster-api-provider-nutanix/test/e2e/csi_test.go:74
Successfully created the nutanix client
STEP: Creating a namespace for hosting the "cluster-csi" test spec 11/17/22 00:29:38.786
INFO: Creating namespace cluster-csi-7xmzcf
INFO: Creating event watcher for namespace "cluster-csi-7xmzcf"
[AfterEach] Nutanix flavor CSI
  /Users/sid.shukla/go/src/github.com/nutanix-cloud-native/cluster-api-provider-nutanix/test/e2e/csi_test.go:97
------------------------------
• [FAILED] [1.115 seconds]
Nutanix flavor CSI [BeforeEach] [capx-feature-test, csi, slow, network]
/Users/sid.shukla/go/src/github.com/nutanix-cloud-native/cluster-api-provider-nutanix/test/e2e/csi_test.go:74
  Create a cluster with Nutanix CSI and use Nutanix Volumes to create PV
  /Users/sid.shukla/go/src/github.com/nutanix-cloud-native/cluster-api-provider-nutanix/test/e2e/csi_test.go:101

  Begin Captured GinkgoWriter Output >>
    [BeforeEach] Nutanix flavor CSI
      /Users/sid.shukla/go/src/github.com/nutanix-cloud-native/cluster-api-provider-nutanix/test/e2e/csi_test.go:74
    STEP: Creating a namespace for hosting the "cluster-csi" test spec 11/17/22 00:29:38.786
    INFO: Creating namespace cluster-csi-7xmzcf
    INFO: Creating event watcher for namespace "cluster-csi-7xmzcf"
    [AfterEach] Nutanix flavor CSI
      /Users/sid.shukla/go/src/github.com/nutanix-cloud-native/cluster-api-provider-nutanix/test/e2e/csi_test.go:97
  << End Captured GinkgoWriter Output

  expected e2econfig variable WEBHOOK_CA to be set
  Expected
      <string>:
  not to be empty
  In [BeforeEach] at: /Users/sid.shukla/go/src/github.com/nutanix-cloud-native/cluster-api-provider-nutanix/test/e2e/test_helpers.go:387

  Full Stack Trace
    github.com/nutanix-cloud-native/cluster-api-provider-nutanix/test/e2e.testHelper.getVariableFromE2eConfig({0xc0020be420?, 0xc00150be00?}, {0x2efde1e, 0xa})
    	/Users/sid.shukla/go/src/github.com/nutanix-cloud-native/cluster-api-provider-nutanix/test/e2e/test_helpers.go:387 +0x20e
    github.com/nutanix-cloud-native/cluster-api-provider-nutanix/test/e2e.glob..func9.1()
    	/Users/sid.shukla/go/src/github.com/nutanix-cloud-native/cluster-api-provider-nutanix/test/e2e/csi_test.go:81 +0x233
```

**Comment by thunderboltsid** [comment]
Projects e2e failed with a timeout
```
• [FAILED] [651.792 seconds]
Nutanix projects [capx-feature-test, projects, slow, network]
/Users/sid.shukla/go/src/github.com/nutanix-cloud-native/cluster-api-provider-nutanix/test/e2e/projects_test.go:40
  [It] Create a cluster linked to an existing project
  /Users/sid.shukla/go/src/github.com/nutanix-cloud-native/cluster-api-provider-nutanix/test/e2e/projects_test.go:123

  Begin Captured GinkgoWriter Output >>
    [BeforeEach] Nutanix projects
      /Users/sid.shukla/go/src/github.com/nutanix-cloud-native/cluster-api-provider-nutanix/test/e2e/projects_test.go:52
    STEP: Creating a namespace for hosting the "cluster-projects" test spec 11/17/22 00:45:02.364
    INFO: Creating namespace cluster-projects-jyucgi
    INFO: Creating event watcher for namespace "cluster-projects-jyucgi"
    [It] Create a cluster linked to an existing project
      /Users/sid.shukla/go/src/github.com/nutanix-cloud-native/cluster-api-provider-nutanix/test/e2e/projects_test.go:123
    STEP: Creating a workload cluster 11/17/22 00:45:02.384
    INFO: Creating the workload cluster with name "cluster-projects-2mgr0d" using the "project" template (Kubernetes v1.23.10, 1 control-plane machines, 1 worker machines)
    INFO: Getting the cluster template yaml
    INFO: clusterctl config cluster cluster-projects-2mgr0d --infrastructure (default) --kubernetes-version v1.23.10 --control-plane-machine-count 1 --worker-machine-count 1 --flavor project
    INFO: Applying the cluster template yaml to the cluster
    INFO: Waiting for the cluster infrastructure to be provisioned
    STEP: Waiting for cluster to enter the provisioned phase 11/17/22 00:45:02.987
    INFO: Waiting for control plane to be initialized
    INFO: Waiting for the first control plane machine managed by cluster-projects-jyucgi/cluster-projects-2mgr0d-kcp to be provisioned
    STEP: Waiting for one control plane node to exist 11/17/22 00:45:43.024
    [AfterEach] Nutanix projects
      /Users/sid.shukla/go/src/github.com/nutanix-cloud-native/cluster-api-provider-nutanix/test/e2e/projects_test.go:61
    STEP: Dumping logs from the "cluster-projects-2mgr0d" workload cluster 11/17/22 00:55:43.004
    STEP: Dumping all the Cluster API resources in the "cluster-projects-jyucgi" namespace 11/17/22 00:55:43.096
    STEP: Deleting cluster cluster-projects-jyucgi/cluster-projects-2mgr0d 11/17/22 00:55:43.467
    STEP: Deleting cluster cluster-projects-2mgr0d 11/17/22 00:55:43.484
    INFO: Waiting for the Cluster cluster-projects-jyucgi/cluster-projects-2mgr0d to be deleted
    STEP: Waiting for cluster cluster-projects-2mgr0d to be deleted 11/17/22 00:55:43.529
    STEP: Deleting namespace used for hosting the "cluster-projects" test spec 11/17/22 00:55:53.544
    INFO: Deleting namespace cluster-projects-jyucgi
  << End Captured GinkgoWriter Output

  Timed out after 600.000s.
  No Control Plane machines came into existence.
  Expected
      <bool>: false
  to be true
  In [It] at: /Users/sid.shukla/go/pkg/mod/sigs.k8s.io/cluster-api/test@v1.2.0-beta.0.0.20220823125924-6e30f66cc3df/framework/controlplane_helpers.go:154

  Full Stack Trace
    sigs.k8s.io/cluster-api/test/framework.WaitForOneKubeadmControlPlaneMachineToExist({0x3290dc8?, 0xc0006a7d80}, {{0xaf98c00?, 0xc000299180?}, 0xc000f151e0?, 0xc00111a000?}, {0xc00222e660, 0x2, 0x2})
    	/Users/sid.shukla/go/pkg/mod/sigs.k8s.io/cluster-api/test@v1.2.0-beta.0.0.20220823125924-6e30f66cc3df/framework/controlplane_helpers.go:154 +0x439
    sigs.k8s.io/cluster-api/test/framework.DiscoveryAndWaitForControlPlaneInitialized({0x3290dc8?, 0xc0006a7d80}, {{0xaf98c00?, 0xc000299180?}, 0xc000f151e0?}, {0xc00222e660, 0x2, 0x2})
    	/Users/sid.shukla/go/pkg/mod/sigs.k8s.io/cluster-api/test@v1.2.0-beta.0.0.20220823125924-6e30f66cc3df/framework/controlplane_helpers.go:249 +0x54b
    sigs.k8s.io/cluster-api/test/framework/clusterctl.setDefaults.func1({_, _}, {{0x329f5f8, 0xc000c593c0}, {{0xc000c52b80, 0x73}, {0xc00135e5b4, 0x82}, {0xc00135e637, 0x43}, ...}, ...}, ...)
    	/Users/sid.shukla/go/pkg/mod/sigs.k8s.io/cluster-api/test@v1.2.0-beta.0.0.20220823125924-6e30f66cc3df/framework/clusterctl/clusterctl_helpers.go:343 +0x9e
    sigs.k8s.io/cluster-api/test/framework/clusterctl.ApplyClusterTemplateAndWait({_, _}, {{0x329f5f8, 0xc000c593c0}, {{0xc000c52b80, 0x73}, {0xc00135e5b4, 0x82}, {0xc00135e637, 0x43}, ...}, ...}, ...)
    	/Users/sid.shukla/go/pkg/mod/sigs.k8s.io/cluster-api/test@v1.2.0-beta.0.0.20220823125924-6e30f66cc3df/framework/clusterctl/clusterctl_helpers.go:304 +0xd10
    github.com/nutanix-cloud-native/cluster-api-provider-nutanix/test/e2e.testHelper.deployClusterAndWait({0xc000116a20?, 0xc00150be00?}, {{0xc000d87578, 0x17}, 0xc000bd18c0, {0x2ef909b, 0x7}, {0xc00135e5b4, 0x82}, {0xc00135e4e0, ...}, ...}, ...)
    	/Users/sid.shukla/go/src/github.com/nutanix-cloud-native/cluster-api-provider-nutanix/test/e2e/test_helpers.go:284 +0x4c5
    github.com/nutanix-cloud-native/cluster-api-provider-nutanix/test/e2e.glob..func20.4()
    	/Users/sid.shukla/go/src/github.com/nutanix-cloud-native/cluster-api-provider-nutanix/test/e2e/projects_test.go:129 +0x216
```

**Comment by thunderboltsid** [comment]
CCM Flavor test failed due to failed assertion on labels
```
• [FAILED] [173.733 seconds]
Nutanix flavor CCM [capx-feature-test, ccm, slow, network]
/Users/sid.shukla/go/src/github.com/nutanix-cloud-native/cluster-api-provider-nutanix/test/e2e/ccm_test.go:33
  [It] Create a cluster with Nutanix CCM
  /Users/sid.shukla/go/src/github.com/nutanix-cloud-native/cluster-api-provider-nutanix/test/e2e/ccm_test.go:68

  Begin Captured GinkgoWriter Output >>
    [BeforeEach] Nutanix flavor CCM
      /Users/sid.shukla/go/src/github.com/nutanix-cloud-native/cluster-api-provider-nutanix/test/e2e/ccm_test.go:51
    STEP: Creating a namespace for hosting the "cluster-ccm" test spec 11/17/22 00:55:54.813
    INFO: Creating namespace cluster-ccm-b75dfw
    INFO: Creating event watcher for namespace "cluster-ccm-b75dfw"
    [It] Create a cluster with Nutanix CCM
      /Users/sid.shukla/go/src/github.com/nutanix-cloud-native/cluster-api-provider-nutanix/test/e2e/ccm_test.go:68
    STEP: Creating a workload cluster 11/17/22 00:55:54.83
    INFO: Creating the workload cluster with name "cluster-ccm-noee0q" using the "ccm" template (Kubernetes v1.23.10, 1 control-plane machines, 1 worker machines)
    INFO: Getting the cluster template yaml
    INFO: clusterctl config cluster cluster-ccm-noee0q --infrastructure (default) --kubernetes-version v1.23.10 --control-plane-machine-count 1 --worker-machine-count 1 --flavor ccm
    INFO: Applying the cluster template yaml to the cluster
    INFO: Waiting for the cluster infrastructure to be provisioned
    STEP: Waiting for cluster to enter the provisioned phase 11/17/22 00:55:55.661
    INFO: Waiting for control plane to be initialized
    INFO: Waiting for the first control plane machine managed by cluster-ccm-b75dfw/cluster-ccm-noee0q-kcp to be provisioned
    STEP: Waiting for one control plane node to exist 11/17/22 00:56:35.704
    INFO: Waiting for control plane to be ready
    INFO: Waiting for control plane cluster-ccm-b75dfw/cluster-ccm-noee0q-kcp to be ready (implies underlying nodes to be ready as well)
    STEP: Waiting for the control plane to be ready 11/17/22 00:57:15.757
    STEP: Checking all the the control plane machines are in the expected failure domains 11/17/22 00:57:45.793
    INFO: Waiting for the machine deployments to be provisioned
    STEP: Waiting for the workload nodes to exist 11/17/22 00:57:45.812
    STEP: Checking all the machines controlled by cluster-ccm-noee0q-wmd are in the "<None>" failure domain 11/17/22 00:58:05.842
    INFO: Waiting for the machine pools to be provisioned
    STEP: Fetching workload proxy 11/17/22 00:58:05.874
    STEP: Checking if nodes have correct CCM labels 11/17/22 00:58:05.903
    [AfterEach] Nutanix flavor CCM
      /Users/sid.shukla/go/src/github.com/nutanix-cloud-native/cluster-api-provider-nutanix/test/e2e/ccm_test.go:64
    STEP: Dumping logs from the "cluster-ccm-noee0q" workload cluster 11/17/22 00:58:06.957
    STEP: Dumping all the Cluster API resources in the "cluster-ccm-b75dfw" namespace 11/17/22 00:58:07.056
    STEP: Deleting cluster cluster-ccm-b75dfw/cluster-ccm-noee0q 11/17/22 00:58:07.219
    STEP: Deleting cluster cluster-ccm-noee0q 11/17/22 00:58:07.237
    INFO: Waiting for the Cluster cluster-ccm-b75dfw/cluster-ccm-noee0q to be deleted
    STEP: Waiting for cluster cluster-ccm-noee0q to be deleted 11/17/22 00:58:07.255
    STEP: Deleting namespace used for hosting the "cluster-ccm" test spec 11/17/22 00:58:47.28
    INFO: Deleting namespace cluster-ccm-b75dfw
  << End Captured GinkgoWriter Output

  Expected
      <string>:
  to match keys: {
  missing expected key node.kubernetes.io/instance-type
  }

  In [It] at: /Users/sid.shukla/go/src/github.com/nutanix-cloud-native/cluster-api-provider-nutanix/test/e2e/ccm_test.go:92

  Full Stack Trace
    github.com/nutanix-cloud-native/cluster-api-provider-nutanix/test/e2e.glob..func3.3()
    	/Users/sid.shukla/go/src/github.com/nutanix-cloud-native/cluster-api-provider-nutanix/test/e2e/ccm_test.go:92 +0x4c8
```

File: templates/ccm/nutanix-ccm.yaml [hunk 3]
```diff
@@ -1,4 +1,12 @@
 ---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: user-ca-bundle
+  namespace: kube-system
+data:
+  ca.crt: NUTANIX_ADDITIONAL_TRUST_BUNDLE
```
Comment (yannickstruyf3) [inline templates/ccm/nutanix-ccm.yaml:8]: Did this pass e2e? Shouldn't this be a variable instead?

File: test/e2e/test_helpers.go [hunk 4]
```diff
@@ -279,6 +279,7 @@ func (t testHelper) deployClusterAndWait(params deployClusterParams, clusterReso
 		ControlPlaneMachineCount: pointer.Int64Ptr(1),
 		WorkerMachineCount:       pointer.Int64Ptr(1),
 	}
+	fmt.Println(cc)
```
Comment (yannickstruyf3) [inline test/e2e/test_helpers.go:282]: Can be removed

File: controllers/nutanixcluster_controller.go [hunk 5]
```diff
@@ -152,10 +154,10 @@ func (r *NutanixClusterReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 	}
 	conditions.MarkTrue(cluster, infrav1.CredentialRefSecretOwnerSetCondition)
 
-	client, err := CreateNutanixClient(ctx, r.SecretInformer, cluster)
+	v3Client, err := CreateNutanixClient(r.SecretInformer, r.ConfigMapInformer, cluster)
```
Comment (yannickstruyf3) [inline controllers/nutanixcluster_controller.go:157]: Don't we need a "reconcileAdditionalTrustBundle" here?

File: pkg/client/client.go [hunk 6]
```diff
@@ -32,6 +29,8 @@ import (
 	nutanixClientV3 "github.com/nutanix-cloud-native/prism-go-client/v3"
 	coreinformers "k8s.io/client-go/informers/core/v1"
 	"k8s.io/klog"
+	"os"
+	"path/filepath"
```
Comment (yannickstruyf3) [inline pkg/client/client.go:33]: Can we move these imports back to the top?

File: templates/ccm/nutanix-ccm.yaml (hunk 3 above)
Comment (thunderboltsid) [reply templates/ccm/nutanix-ccm.yaml:8]: See the `cluster-templates` makefile target; there's a `sed` command that replaces that.

File: templates/base/cm.yaml [hunk 7]
```diff
@@ -0,0 +1,8 @@
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: user-ca-bundle
+  namespace: "${NAMESPACE}"
+data:
+  ca.crt: NUTANIX_ADDITIONAL_TRUST_BUNDLE
```
[the remark this answers is not in the document]
Comment (thunderboltsid) [reply templates/base/cm.yaml:8]: See the `cluster-templates` makefile target; there's a `sed` command that replaces this after kustomize runs.

File: config/crd/bases/infrastructure.cluster.x-k8s.io_nutanixclusters.yaml [hunk 8]
```diff
@@ -245,6 +270,31 @@ spec:
...
+                      trusted RootCA. The trust bundle is added to the cert pool used
+                      to authenticate the TLS connection to the Prism Central.
+                    properties:
+                      data:
+                        description: Data of the trust bundle if Kind is String.
+                        type: string
+                      kind:
+                        description: Kind of the Nutanix trust bundle
+                        enum:
+                        - String
```
Comment (deepakm-ntnx) [inline config/crd/bases/infrastructure.cluster.x-k8s.io_nutanixclusters.yaml:286]: why do we need string or configmap? cant it be always string?

File: controllers/nutanixcluster_controller.go (hunk 5 above)
Comment (deepakm-ntnx) [reply controllers/nutanixcluster_controller.go:157]: just checking, do we want to support users to update this additionalbundle later

Comment (thunderboltsid) [reply controllers/nutanixcluster_controller.go:157]: The trust bundle only needs to be updated when the root CA used to issue the certificate has changed. Wondering if it's often enough to warrant handling with a reconciler.

File: config/crd/bases/infrastructure.cluster.x-k8s.io_nutanixclusters.yaml (hunk 8 above)
Comment (thunderboltsid) [reply config/crd/bases/infrastructure.cluster.x-k8s.io_nutanixclusters.yaml:286]: The String type is for local testing / dev purposes. When it's running on a kubernetes cluster, it relies on the config map.

File: config/crd/bases/infrastructure.cluster.x-k8s.io_nutanixclusters.yaml [hunk 9]
```diff
@@ -71,6 +71,31 @@ spec:
...
```
[the remark this answers is not in the document]
Comment (thunderboltsid) [reply config/crd/bases/infrastructure.cluster.x-k8s.io_nutanixclusters.yaml:87]: duplicate comment; see reply above.

**Review by deepakm-ntnx — DISMISSED** [review]

File: controllers/nutanixcluster_controller.go (hunk 5 above)
Comment (deepakm-ntnx) [reply controllers/nutanixcluster_controller.go:157]: lets track it separately

**Comment by thunderboltsid** [comment]
@yannickstruyf3 we need to create a CCM version with the updated `v0.3.4` Prism Go Client.

File: Makefile [hunk 10]
```diff
@@ -263,6 +264,11 @@ undeploy: ## Undeploy controller from the K8s cluster specified in ~/.kube/confi
 .PHONY: cluster-e2e-templates
 cluster-e2e-templates: $(KUSTOMIZE) cluster-e2e-templates-v1beta1 cluster-e2e-templates-v1alpha4 ## Generate cluster templates for all versions
 
+# `NUTANIX_ADDITIONAL_TRUST_BUNDLE` is substituted for `"${NUTANIX_ADDITIONAL_TRUST_BUNDLE=''}"` after generating the templates
+# using `sed`. This ensures the double quotes are preserved around the environment variable. When envsubst is called
+# by clusterctl downstream to substitute the variable with a one-liner certificate with embedded `\n` characters (generated
+# using `awk 'NF {sub(/\r/, ""); printf "%s\\n",$0;}' ca-cert.pem`), the resulting template honors the escaped newlines.
```
Comment (yannickstruyf3) [inline Makefile:270]: Is this still relevant?

Comment (thunderboltsid) [reply Makefile:270]: Good catch; forgot to push that commit

**Review by yannickstruyf3 — APPROVED** [review]

**Pull request merged**